File: experiments/adrianco/experiment-5/runs/language=clojure_model=claude-opus-4-8_tooling=none/rep2/src/brazilian_soccer/mcp.clj
;; =============================================================================
;; brazilian-soccer.mcp
;; -----------------------------------------------------------------------------
;; CONTEXT
;;   Entry point for the Brazilian Soccer MCP server (see TASK.md). Implements
;;   the Model Context Protocol (https://modelcontextprotocol.io) stdio
;;   transport: newline-delimited JSON-RPC 2.0 messages on stdin/stdout, with
;;   all human-facing logging directed to stderr so it never corrupts the
;;   protocol stream.
;;
;;   Supported JSON-RPC methods:
;;     initialize, notifications/initialized, ping, tools/list, tools/call
;;
;;   Each entry in `tools` declares an MCP tool (name, description, JSON-Schema
;;   for its arguments) and a handler. Handlers call brazilian-soccer.queries,
;;   then return BOTH a formatted prose block (brazilian-soccer.format) and the
;;   raw structured data as JSON, so an LLM client can present prose or reason
;;   over the structure. Tool coverage maps onto the spec's five capability
;;   categories (match / team / player / competition / statistical queries).
;;
;;   Run with:  clojure -M:run        (speaks MCP over stdio)
;; =============================================================================
(ns brazilian-soccer.mcp
  (:require [brazilian-soccer.data :as data]
            [brazilian-soccer.format :as fmt]
            [brazilian-soccer.queries :as q]
            [clojure.data.json :as json]
            [clojure.string :as str])
  (:gen-class))

(def protocol-version "2024-11-05")
(def server-info {:name "brazilian-soccer-mcp" :version "1.0.0"})

(defn- log [& xs]
  (binding [*out* *err*]
    (apply println xs)
    (flush)))

;; -----------------------------------------------------------------------------
;; Tool handlers — each takes a keyword-keyed argument map and returns
;; {:text <prose> :data <edn>}.
;; -----------------------------------------------------------------------------

(defn- h-search-matches [a]
  (let [ms (q/search-matches (select-keys a [:team :home :away :opponent
                                             :competition :season :from :to :limit]))]
    {:text (str (count ms) " match(es) found:\n" (fmt/matches->text ms))
     :data ms}))

(defn- h-head-to-head [a]
  (let [r (q/head-to-head (:team_a a) (:team_b a))]
    {:text (fmt/head-to-head->text r) :data (dissoc r :matches)}))

(defn- h-team-stats [a]
  (let [venue (case (some-> (:venue a) str/lower-case)
                "home" :home "away" :away :all)
        r (q/team-stats (:team a)
                        (-> (select-keys a [:competition :season :from :to])
                            (assoc :venue venue)))]
    {:text (fmt/team-stats->text r) :data r}))

(defn- h-standings [a]
  (let [rows (q/standings (:competition a) (:season a))]
    {:text (fmt/standings->text rows a) :data rows}))

(defn- h-competition-stats [a]
  (let [r (q/competition-stats (select-keys a [:competition :season :team :from :to]))]
    {:text (fmt/competition-stats->text r) :data r}))

(defn- h-biggest-wins [a]
  (let [ms (q/biggest-wins (select-keys a [:competition :season :team :limit]))]
    {:text (fmt/biggest-wins->text ms) :data ms}))

(defn- h-search-players [a]
  (let [ps (q/search-players (-> (select-keys a [:name :nationality :club :position :limit])
                                 (assoc :min-overall (:min_overall a))))]
    {:text (str (count ps) " player(s) found:\n" (fmt/players->text ps)) :data ps}))

(defn- h-top-players [a]
  (let [ps (q/top-players (select-keys a [:nationality :club :position :limit]))]
    {:text (fmt/players->text ps) :data ps}))

(defn- h-list-competitions [_]
  (let [ms (data/all-matches)
        comps (->> ms (map :competition) frequencies
                   (sort-by (comp - val)))
        seasons (->> ms (keep :season) distinct sort vec)]
    {:text (str "Competitions (with match counts):\n"
                (str/join "\n" (map (fn [[c n]] (format "- %s: %d matches" c n)) comps))
                (format "\n\nSeasons covered: %s–%s"
                        (first seasons) (last seasons)))
     :data {:competitions (mapv (fn [[c n]] {:competition c :matches n}) comps)
            :seasons seasons
            :total-matches (count ms)
            :total-players (count (data/all-players))}}))

;; -----------------------------------------------------------------------------
;; Tool registry
;; -----------------------------------------------------------------------------

(def ^:private str-schema  {:type "string"})
(def ^:private int-schema  {:type "integer"})

(def tools
  [{:name "search_matches"
    :description "Search matches across all competitions by team, opponent, home/away, competition, season, or date range. Returns matches most-recent first."
    :inputSchema {:type "object"
                  :properties {:team        (assoc str-schema :description "Team involved as home or away (fuzzy, accent-insensitive)")
                               :opponent    (assoc str-schema :description "With `team`, narrows to matches between the two teams")
                               :home        (assoc str-schema :description "Team as home side only")
                               :away        (assoc str-schema :description "Team as away side only")
                               :competition (assoc str-schema :description "Competition name substring, e.g. 'Brasileirão', 'Copa do Brasil', 'Libertadores'")
                               :season      (assoc int-schema :description "Season year, e.g. 2019")
                               :from        (assoc str-schema :description "Earliest match date, ISO YYYY-MM-DD")
                               :to          (assoc str-schema :description "Latest match date, ISO YYYY-MM-DD")
                               :limit       (assoc int-schema :description "Max results (default 100)")}}
    :handler h-search-matches}

   {:name "head_to_head"
    :description "Head-to-head record and match list between two teams across all competitions."
    :inputSchema {:type "object"
                  :properties {:team_a (assoc str-schema :description "First team")
                               :team_b (assoc str-schema :description "Second team")}
                  :required ["team_a" "team_b"]}
    :handler h-head-to-head}

   {:name "team_stats"
    :description "Win/draw/loss and goals record for a team, optionally restricted by competition, season, date range and venue (home/away/all)."
    :inputSchema {:type "object"
                  :properties {:team        (assoc str-schema :description "Team name")
                               :competition (assoc str-schema :description "Competition name substring")
                               :season      (assoc int-schema :description "Season year")
                               :venue       (assoc str-schema :description "'home', 'away' or 'all' (default all)")
                               :from        (assoc str-schema :description "Earliest date ISO")
                               :to          (assoc str-schema :description "Latest date ISO")}
                  :required ["team"]}
    :handler h-team-stats}

   {:name "standings"
    :description "League table for a competition and season, computed from match results (3 pts win, 1 draw). Best for Brasileirão Série A."
    :inputSchema {:type "object"
                  :properties {:competition (assoc str-schema :description "Competition name, e.g. 'Brasileirão Série A'")
                               :season      (assoc int-schema :description "Season year, e.g. 2019")
                               :limit       (assoc int-schema :description "Number of table rows to show (default 20)")}
                  :required ["competition" "season"]}
    :handler h-standings}

   {:name "competition_stats"
    :description "Aggregate statistics (match count, total/average goals, home/away/draw rates) over a filtered match set."
    :inputSchema {:type "object"
                  :properties {:competition (assoc str-schema :description "Competition name substring")
                               :season      (assoc int-schema :description "Season year")
                               :team        (assoc str-schema :description "Restrict to a team's matches")
                               :from        (assoc str-schema :description "Earliest date ISO")
                               :to          (assoc str-schema :description "Latest date ISO")}}
    :handler h-competition-stats}

   {:name "biggest_wins"
    :description "Matches with the largest goal margins, optionally filtered by competition, season or team."
    :inputSchema {:type "object"
                  :properties {:competition (assoc str-schema :description "Competition name substring")
                               :season      (assoc int-schema :description "Season year")
                               :team        (assoc str-schema :description "Restrict to a team's matches")
                               :limit       (assoc int-schema :description "Max results (default 10)")}}
    :handler h-biggest-wins}

   {:name "search_players"
    :description "Search the FIFA player database by name, nationality, club, position and/or minimum overall rating. Sorted by rating."
    :inputSchema {:type "object"
                  :properties {:name        (assoc str-schema :description "Player name substring")
                               :nationality (assoc str-schema :description "Nationality, e.g. 'Brazil'")
                               :club        (assoc str-schema :description "Club name substring, e.g. 'Flamengo'")
                               :position    (assoc str-schema :description "Position code, e.g. 'ST', 'GK', 'LW'")
                               :min_overall (assoc int-schema :description "Minimum FIFA overall rating")
                               :limit       (assoc int-schema :description "Max results (default 50)")}}
    :handler h-search-players}

   {:name "top_players"
    :description "Highest-rated players, optionally filtered by nationality, club or position. E.g. top Brazilian players."
    :inputSchema {:type "object"
                  :properties {:nationality (assoc str-schema :description "Nationality filter, e.g. 'Brazil'")
                               :club        (assoc str-schema :description "Club filter")
                               :position    (assoc str-schema :description "Position filter")
                               :limit       (assoc int-schema :description "Max results (default 10)")}}
    :handler h-top-players}

   {:name "list_competitions"
    :description "List the competitions and seasons available in the dataset, with match and player counts. Good first call to discover coverage."
    :inputSchema {:type "object" :properties {}}
    :handler h-list-competitions}])

(def ^:private tool-by-name
  (into {} (map (juxt :name identity) tools)))

(defn- tool-defs
  "Tool list for tools/list (handlers stripped)."
  []
  (mapv #(dissoc % :handler) tools))

;; -----------------------------------------------------------------------------
;; JSON-RPC dispatch
;; -----------------------------------------------------------------------------

(defn- call-tool [name arguments]
  (if-let [{:keys [handler]} (tool-by-name name)]
    (let [{:keys [text data]} (handler (or arguments {}))]
      {:content [{:type "text" :text text}
                 {:type "text"
                  :text (str "```json\n"
                             (json/write-str data {:escape-unicode false})
                             "\n```")}]
       :isError false})
    {:content [{:type "text" :text (str "Unknown tool: " name)}]
     :isError true}))

(defn handle-request
  "Pure dispatch from a parsed JSON-RPC request map (keyword keys) to a result
   value or {:error {...}}. Returns nil for notifications (no response)."
  [{:keys [method params id]}]
  (case method
    "initialize"
    {:protocolVersion protocol-version
     :capabilities {:tools {}}
     :serverInfo server-info}

    ("notifications/initialized" "initialized" "notifications/cancelled")
    nil

    "ping" {}

    "tools/list" {:tools (tool-defs)}

    "tools/call"
    (try
      (call-tool (:name params) (:arguments params))
      (catch Exception e
        (log "tool error:" (.getMessage e))
        {:content [{:type "text" :text (str "Error: " (.getMessage e))}]
         :isError true}))

    ;; unknown method
    (if id
      {:error {:code -32601 :message (str "Method not found: " method)}}
      nil)))

(defn- response-for [req]
  (let [{:keys [id method]} req]
    (when (some? id) ; requests (with id) get a response; notifications don't
      (let [result (handle-request req)]
        (cond
          (and (map? result) (:error result))
          {:jsonrpc "2.0" :id id :error (:error result)}
          :else
          {:jsonrpc "2.0" :id id :result result})))))

;; -----------------------------------------------------------------------------
;; stdio loop
;; -----------------------------------------------------------------------------

(defn- write-message! [writer msg]
  (locking writer
    (.write writer ^String (json/write-str msg {:escape-unicode false}))
    (.write writer "\n")
    (.flush writer)))

(defn serve!
  "Read newline-delimited JSON-RPC requests from `reader`, write responses to
   `writer`. Blocks until EOF."
  [reader writer]
  (loop []
    (when-let [line (.readLine ^java.io.BufferedReader reader)]
      (when-not (str/blank? line)
        (let [req (try (json/read-str line {:key-fn keyword})
                       (catch Exception e
                         (log "parse error:" (.getMessage e)) nil))]
          (when req
            (try
              (when-let [resp (response-for req)]
                (write-message! writer resp))
              (catch Exception e
                (log "dispatch error:" (.getMessage e))
                (when-let [id (:id req)]
                  (write-message! writer
                                  {:jsonrpc "2.0" :id id
                                   :error {:code -32603 :message (.getMessage e)}})))))))
      (recur))))

(defn -main [& _]
  (log "Brazilian Soccer MCP server starting"
       (format "(%d matches, %d players)"
               (count (data/all-matches)) (count (data/all-players))))
  (let [reader (java.io.BufferedReader. *in*)
        writer *out*]
    (serve! reader writer)
    (log "Brazilian Soccer MCP server stopped")))

File: experiments/adrianco/experiment-7/brazil/runs/language=rust_model=claude-opus-4-8-fast/rep2/src/mcp.rs
// =============================================================================
// Context
// -----------------------------------------------------------------------------
// Module:  mcp
// Purpose: Model Context Protocol server implementation. Speaks JSON-RPC 2.0
//          over newline-delimited stdio (the MCP "stdio" transport). It
//          advertises eight tools - one per query-engine capability - and
//          dispatches `tools/call` requests onto `queries.rs`, returning the
//          formatted text as MCP `content` blocks.
//
//          The protocol handling is split out from I/O so it can be unit
//          tested: `handle_request` is a pure (Dataset, Value) -> Option<Value>
//          function, and `serve_stdio` is the thin runtime loop in main.rs.
//
// Methods: initialize, notifications/initialized, tools/list, tools/call.
// =============================================================================

use crate::data::Dataset;
use crate::queries::MatchFilter;
use serde_json::{json, Value};

const PROTOCOL_VERSION: &str = "2024-11-05";
const DEFAULT_MATCH_LIMIT: usize = 25;
const DEFAULT_WINS_LIMIT: usize = 10;
const DEFAULT_PLAYER_LIMIT: usize = 25;

/// JSON-RPC error codes used by this server.
mod codes {
    pub const METHOD_NOT_FOUND: i64 = -32601;
    pub const INVALID_PARAMS: i64 = -32602;
}

/// The list of tools advertised to clients, as an MCP `tools/list` payload.
pub fn tool_definitions() -> Value {
    json!([
        {
            "name": "search_matches",
            "description": "Search matches by team, opponent, competition and/or season. When both 'team' and 'opponent' are given, a head-to-head summary is appended. Results are newest-first.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team": {"type": "string", "description": "Team name, e.g. 'Flamengo' (state suffixes and accents optional)"},
                    "opponent": {"type": "string", "description": "Optional second team to restrict to matches between the two"},
                    "competition": {"type": "string", "description": "Optional competition filter, e.g. 'Brasileirão', 'Copa do Brasil', 'Libertadores'"},
                    "season": {"type": "integer", "description": "Optional season year, e.g. 2019"},
                    "limit": {"type": "integer", "description": "Max matches to list (default 25)"}
                }
            }
        },
        {
            "name": "team_stats",
            "description": "Win/draw/loss record, goals for/against and win rate for a team, optionally restricted by season, competition and venue.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team": {"type": "string"},
                    "season": {"type": "integer"},
                    "competition": {"type": "string"},
                    "venue": {"type": "string", "enum": ["home", "away", "all"], "description": "Restrict to home or away matches (default all)"}
                },
                "required": ["team"]
            }
        },
        {
            "name": "head_to_head",
            "description": "Head-to-head record (wins, draws, goals) between two teams across all competitions in the dataset.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team1": {"type": "string"},
                    "team2": {"type": "string"}
                },
                "required": ["team1", "team2"]
            }
        },
        {
            "name": "competition_standings",
            "description": "League standings for a competition and season, calculated from match results (3 points per win, 1 per draw).",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "competition": {"type": "string", "description": "e.g. 'Brasileirão'"},
                    "season": {"type": "integer"}
                },
                "required": ["competition", "season"]
            }
        },
        {
            "name": "league_statistics",
            "description": "Aggregate statistics (average goals per match, home/away win rates, draw rate) for a competition and/or season, or the whole dataset.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "competition": {"type": "string"},
                    "season": {"type": "integer"}
                }
            }
        },
        {
            "name": "biggest_wins",
            "description": "The largest-margin victories matching an optional team / competition / season filter.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team": {"type": "string"},
                    "competition": {"type": "string"},
                    "season": {"type": "integer"},
                    "limit": {"type": "integer", "description": "default 10"}
                }
            }
        },
        {
            "name": "list_competitions",
            "description": "List all competitions in the dataset with match counts and the range of seasons covered.",
            "inputSchema": {"type": "object", "properties": {}}
        },
        {
            "name": "search_players",
            "description": "Search FIFA player data by name, nationality, club and/or position. Results sorted by overall rating, highest first.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "name": {"type": "string", "description": "Substring of the player's name"},
                    "nationality": {"type": "string", "description": "e.g. 'Brazil'"},
                    "club": {"type": "string", "description": "e.g. 'Flamengo'"},
                    "position": {"type": "string", "description": "e.g. 'GK', 'ST', 'CB'"},
                    "limit": {"type": "integer", "description": "default 25"}
                }
            }
        }
    ])
}

fn str_arg<'a>(args: &'a Value, key: &str) -> Option<&'a str> {
    args.get(key).and_then(|v| v.as_str()).filter(|s| !s.is_empty())
}

fn int_arg(args: &Value, key: &str) -> Option<i32> {
    args.get(key).and_then(|v| match v {
        Value::Number(n) => n.as_i64().map(|i| i as i32),
        Value::String(s) => s.parse().ok(),
        _ => None,
    })
}

fn usize_arg(args: &Value, key: &str, default: usize) -> usize {
    int_arg(args, key)
        .filter(|&n| n > 0)
        .map(|n| n as usize)
        .unwrap_or(default)
}

/// Run a single tool call, returning either the formatted text or an error
/// message describing a missing required argument.
pub fn dispatch_tool(ds: &Dataset, name: &str, args: &Value) -> Result<String, String> {
    match name {
        "search_matches" => {
            let f = MatchFilter {
                team: str_arg(args, "team").map(String::from),
                team2: str_arg(args, "opponent").map(String::from),
                competition: str_arg(args, "competition").map(String::from),
                season: int_arg(args, "season"),
            };
            Ok(ds.search_matches(&f, usize_arg(args, "limit", DEFAULT_MATCH_LIMIT)))
        }
        "team_stats" => {
            let team = str_arg(args, "team").ok_or("missing required argument 'team'")?;
            Ok(ds.team_stats(
                team,
                int_arg(args, "season"),
                str_arg(args, "competition"),
                str_arg(args, "venue").unwrap_or("all"),
            ))
        }
        "head_to_head" => {
            let t1 = str_arg(args, "team1").ok_or("missing required argument 'team1'")?;
            let t2 = str_arg(args, "team2").ok_or("missing required argument 'team2'")?;
            Ok(ds.head_to_head(t1, t2))
        }
        "competition_standings" => {
            let comp =
                str_arg(args, "competition").ok_or("missing required argument 'competition'")?;
            let season = int_arg(args, "season").ok_or("missing required argument 'season'")?;
            Ok(ds.standings(comp, season))
        }
        "league_statistics" => Ok(ds.league_statistics(
            str_arg(args, "competition"),
            int_arg(args, "season"),
        )),
        "biggest_wins" => {
            let f = MatchFilter {
                team: str_arg(args, "team").map(String::from),
                team2: None,
                competition: str_arg(args, "competition").map(String::from),
                season: int_arg(args, "season"),
            };
            Ok(ds.biggest_wins(&f, usize_arg(args, "limit", DEFAULT_WINS_LIMIT)))
        }
        "list_competitions" => Ok(ds.list_competitions()),
        "search_players" => Ok(ds.search_players(
            str_arg(args, "name"),
            str_arg(args, "nationality"),
            str_arg(args, "club"),
            str_arg(args, "position"),
            usize_arg(args, "limit", DEFAULT_PLAYER_LIMIT),
        )),
        other => Err(format!("unknown tool '{other}'")),
    }
}

fn result(id: Value, result: Value) -> Value {
    json!({"jsonrpc": "2.0", "id": id, "result": result})
}

fn error(id: Value, code: i64, message: &str) -> Value {
    json!({"jsonrpc": "2.0", "id": id, "error": {"code": code, "message": message}})
}

/// Handle one parsed JSON-RPC request. Returns `Some(response)` for requests
/// and `None` for notifications (which must not be answered).
pub fn handle_request(ds: &Dataset, req: &Value) -> Option<Value> {
    let method = req.get("method").and_then(|m| m.as_str()).unwrap_or("");
    let id = req.get("id").cloned();

    // Notifications carry no id and expect no response.
    let id = id?;

    match method {
        "initialize" => Some(result(
            id,
            json!({
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": {"tools": {}},
                "serverInfo": {
                    "name": "brazilian-soccer-mcp",
                    "version": env!("CARGO_PKG_VERSION")
                }
            }),
        )),
        "ping" => Some(result(id, json!({}))),
        "tools/list" => Some(result(id, json!({"tools": tool_definitions()}))),
        "tools/call" => {
            let params = req.get("params").cloned().unwrap_or(json!({}));
            let Some(name) = params.get("name").and_then(|n| n.as_str()) else {
                return Some(error(id, codes::INVALID_PARAMS, "missing tool name"));
            };
            let args = params.get("arguments").cloned().unwrap_or(json!({}));
            match dispatch_tool(ds, name, &args) {
                Ok(text) => Some(result(
                    id,
                    json!({"content": [{"type": "text", "text": text}], "isError": false}),
                )),
                Err(msg) => Some(result(
                    id,
                    json!({"content": [{"type": "text", "text": msg}], "isError": true}),
                )),
            }
        }
        other => Some(error(
            id,
            codes::METHOD_NOT_FOUND,
            &format!("method not found: {other}"),
        )),
    }
}

/// Runtime loop: read newline-delimited JSON-RPC messages from `input`, write
/// responses to `output`. Blocks until EOF on `input`.
pub fn serve_stdio<R: std::io::BufRead, W: std::io::Write>(
    ds: &Dataset,
    mut input: R,
    mut output: W,
) -> std::io::Result<()> {
    let mut line = String::new();
    loop {
        line.clear();
        let n = input.read_line(&mut line)?;
        if n == 0 {
            break; // EOF
        }
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let response = match serde_json::from_str::<Value>(trimmed) {
            Ok(req) => handle_request(ds, &req),
            Err(e) => Some(json!({
                "jsonrpc": "2.0",
                "id": null,
                "error": {"code": -32700, "message": format!("parse error: {e}")}
            })),
        };
        if let Some(resp) = response {
            writeln!(output, "{}", serde_json::to_string(&resp)?)?;
            output.flush()?;
        }
    }
    Ok(())
}
